src/libsystemd-network/dhcp-static-lease.c

```diff
@@ -0,0 +1,95 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include "alloc-util.h"
+#include "dhcp-server-internal.h"
+#include "memory-util.h"
+#include "strv.h"
+
+static sd_dhcp_static_lease *sd_dhcp_static_lease_free(sd_dhcp_static_lease *i) {
+        if (!i)
+                return NULL;
+
+        free(i->client_id.data);
+        return mfree(i);
+}
+
+int sd_dhcp_static_lease_set_client_id_by_mac(
+                sd_dhcp_static_lease *lease,
+                const uint8_t *mac_addr,
+                size_t mac_addr_size) {
+        uint8_t *data;
+
+        assert_return(lease, -EINVAL);
+        assert_return(mac_addr, -EINVAL);
+        assert_return(mac_addr_size == ETH_ALEN, -EINVAL);
+
+        data = new(uint8_t, mac_addr_size + 1);
+        if (!data)
+                return -ENOMEM;
+
+        /* set client id type to 1: Ethernet Link-Layer (RFC 2132) */
+        data[0] = 0x01;
+        memcpy(data + 1, mac_addr, mac_addr_size);
+
+        free_and_replace(lease->client_id.data, data);
+
+        lease->client_id.length = mac_addr_size + 1;
+
+        return 0;
+}
+
+int sd_dhcp_static_lease_unset_client_id(sd_dhcp_static_lease *lease) {
+        assert_return(lease, -EINVAL);
+
+        lease->client_id.data = mfree(lease->client_id.data);
+        lease->client_id.length = 0;
+
+        return 0;
+}
+
+int sd_dhcp_static_lease_set_address(sd_dhcp_static_lease *lease, const struct in_addr *address) {
+        assert_return(lease, -EINVAL);
+        assert_return(address, -EINVAL);
```

Comment:
maybe add:

```c
assert_return(address->s_addr != 0, -EINVAL);
```

here, to make clear that setting and unsetting are two different things, after all 0 is used as marker for unsetting.